### PR TITLE
Add `link` to filehandler and its associated test

### PR DESCRIPTION
**Pull request by aerorahul**
**Description**
This PR:
- adds the ability to `link` files via `FileHandler`
- adds associated test for `link`.
- removes older python version tests as they are deprecated in the GH runner.

**Type of change**
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

**How Has This Been Tested?**

- [X] pynorms
- [X] pytests

**Checklist**

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

File: tests/test_file_utils.py [hunk 1]
```diff
@@ -104,3 +104,57 @@ def test_copy(tmp_path):
...
+    bad_link_list = [[output_dir_path / 'bad_link.txt', input_dir_path / 'non_existent.txt']]
+
+    # Create a config dictionary for FileHandler
+    bad_config = {'link': bad_link_list}
+    FileHandler(bad_config).sync()
+
+    # Follow the bad link to the file and check this is a dead link to a file that does not exist
+    pp = os.path.realpath(output_dir_path / 'bad_link.txt')
+    with pytest.raises(AssertionError):
+        assert os.path.isfile(pp)
```
Comment (DavidHuber-NOAA) [inline tests/test_file_utils.py]:
```suggestion
    assert not os.path.isfile(pp)
```

Comment (aerorahul) [reply tests/test_file_utils.py]: The test is to raise the assertion error on a dead link, asserting that infact the source file does not exist. 
What am I missing?

Comment (DavidHuber-NOAA) [reply tests/test_file_utils.py]: `AssertError` could (at least in theory) be raised for some other reason than `pp` not being a file.  Asserting instead that `pp` is not a file is cleaner and prevents any weird edge cases.

File: src/wxflow/file_utils.py [hunk 2]
```diff
@@ -110,3 +112,27 @@ def _make_dirs(dirlist):
...
+        """
+        for sublist in filelist:
+            if len(sublist) != 2:
+                raise IndexError(
+                    f"List must be of the form ['link name', 'target'], not {sublist}")
+            link_name = sublist[0]
+            target = sublist[1]
+            logger.info(f"Linking {link_name} to {target}")
+            if not os.path.exists(target):
+                logger.warning(f"WARNING: Target file '{target}' does not exist!")
```
Comment (WalterKolczynski-NOAA) [inline src/wxflow/file_utils.py]: Should we add a third `must_exist` flag which elevates this to an Error (maybe a future PR)?

File: tests/test_file_utils.py [hunk 3]
```diff
@@ -104,3 +104,57 @@ def test_copy(tmp_path):
...
+def test_link_files(tmp_path):
+    """
+    Test for linking files:
+    Parameters
+    ----------
+    tmp_path - pytest fixture
+    """
+
+    # Test 1 (nominal operation) - Creating a directory and linking files to it
+    input_dir_path = tmp_path / 'my_input_dir'
```
Comment (WalterKolczynski-NOAA) [inline tests/test_file_utils.py:118]:
```suggestion
    input_dir_path = os.path.join(tmp_path, 'my_input_dir')
```

File: tests/test_file_utils.py [hunk 4]
```diff
@@ -104,3 +104,57 @@ def test_copy(tmp_path):
...
+
+    # Create the input directory
+    config = {'mkdir': [input_dir_path]}
+    FileHandler(config).sync()
+
+    # Put empty files in input_dir_path
+    src_files = [input_dir_path / 'a.txt', input_dir_path / 'b.txt']
```
Comment (WalterKolczynski-NOAA) [inline tests/test_file_utils.py:125]:
```suggestion
    src_files = [os.path.join(input_dir_path, 'a.txt'), os.path.join(input_dir_path, 'b.txt')]
```

Comment (WalterKolczynski-NOAA) [reply tests/test_file_utils.py:125]: etc. for remaining paths

File: src/wxflow/file_utils.py [hunk 5]
```diff
@@ -110,3 +112,27 @@ def _make_dirs(dirlist):
...
+    def _link_files(filelist):
+        """Function to link all files specified in the list
+
+        `filelist` should be in the form:
+        - [link name, target]
+
+        Parameters
+        ----------
+        filelist : list
+                List of lists of [link name, target]
```
Comment (CoryMartin-NOAA) [inline src/wxflow/file_utils.py]: Does target here mean the first half of `ln` as in `ln target dest`? I think that could be confusing since, to me, the target is where the link is pointing to.

File: src/wxflow/file_utils.py [hunk 6]
```diff
@@ -110,3 +112,27 @@ def _make_dirs(dirlist):
...
+        """
+        for sublist in filelist:
+            if len(sublist) != 2:
+                raise IndexError(
```
Comment (WalterKolczynski-NOAA) [inline src/wxflow/file_utils.py:130]: Need a test where there are an incorrect number of args on a line for code coverage (maybe two, one or not enough, one for too many).

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: src/wxflow/file_utils.py (hunk 5 above)
Comment (CoryMartin-NOAA) [reply src/wxflow/file_utils.py]: Yeah so looking at the pathlib documentation, I think we should be a bit more descriptive here, since many people might blindly think the order is:
`[ target, link_name ]` when it is the opposite here because they are used to how `ln` works

Comment (aerorahul) [reply src/wxflow/file_utils.py]: So, the signature of `ln` is as follows:
```bash
ln TARGET LINK_NAME
```
The signature of `cp` is:
```bash
cp SOURCE DESTINATION
```
Requirement of  `copy` list in `wxflow` is `[SOURCE, DESTINATION]`

I wrote `link` with the intention to use the list provided by `copy`.  This will allow us to write the yamls as if targeted for ops (using copy), but then replacing the action `copy` with` link` for research (if needed).

I could, if the desire is to switch the implementation in `link` to the correct signature, we will need a method to `map` existing `copy` list into the correct `link` list without having to duplicate the section in the yaml.

Both are easy enough to do.

Comment (aerorahul) [reply src/wxflow/file_utils.py]: As it stands, `copy` cannot be used as is because `copy` allows the `DESTINATION` to end with `/`; meaning, the `DESTINATION` is a directory and the destination filename is the same as `SOURCE` filename.
I will switch the `link` to the correct form and for now let users write the section of `link` yaml as syntax would want it.

Comment (CoryMartin-NOAA) [reply src/wxflow/file_utils.py]: I'm confused, if you wrote link with that intention (which makes sense) then why would the list order be opposite between FileHandler link and copy?

File: tests/test_file_utils.py (hunk 3 above)
Comment (aerorahul) [reply tests/test_file_utils.py:118]: `tmp_path` is a special fixture object. Its use case is described [here](https://docs.pytest.org/en/stable/how-to/tmp_path.html).  It does not need `os.path.join` and will return a valid path object.  If you still want me to turn this to `os.path.join`, I will do it here and everywhere else I may have used it.

File: src/wxflow/file_utils.py (hunk 5 above)
Comment (aerorahul) [reply src/wxflow/file_utils.py]: Thanks for clearing up my confusion.  I will correct the mistake.

File: tests/test_file_utils.py (hunk 3 above)
Comment (WalterKolczynski-NOAA) [reply tests/test_file_utils.py:118]: Nope, it's fine. Didn't know about that.

File: src/wxflow/file_utils.py (hunk 5 above)
Comment (aerorahul) [reply src/wxflow/file_utils.py]: Thanks @CoryMartin-NOAA. I have applied the suggestions via 2e16ce2

File: src/wxflow/file_utils.py (hunk 2 above)
Comment (aerorahul) [reply src/wxflow/file_utils.py]: I updated the warning to alert a dead link will be created.  
We can employ another method e.g. `link_req` similar to `copy_req` action to ensure the target file exists.

File: src/wxflow/file_utils.py [hunk 7]
```diff
@@ -110,3 +112,29 @@ def _make_dirs(dirlist):
...
+                    f"List must be of the form ['target', 'link name'], not {sublist}")
+            target = sublist[0]
+            link_name = sublist[1]
+            if os.path.isdir(link_name):
+                link_name = os.path.join(link_name, os.path.basename(target))
+            logger.info(f"Linking {link_name} to {target}")
+            if not os.path.exists(target):
+                logger.warning(f"WARNING: Target file '{target}' does not exist, will result in dead link!")
+            Path(link_name).symlink_to(target)
+            logger.info(f'Linked {target} to {link_name}')
```
Comment (CoryMartin-NOAA) [inline src/wxflow/file_utils.py:139]: Do we really need the log to say:
```
Linking foo to bar
Linked foo to bar
```
Seems like just the second one would be fine?

**Review by CoryMartin-NOAA — COMMENTED** [review]
One comment, otherwise I think this is fine

File: src/wxflow/file_utils.py (hunk 6 above)
Comment (aerorahul) [reply src/wxflow/file_utils.py:130]: Test added.

File: src/wxflow/file_utils.py (hunk 7 above)
Comment (aerorahul) [reply src/wxflow/file_utils.py:139]: done.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Review by CoryMartin-NOAA — APPROVED** [review]

**Review by DavidHuber-NOAA — APPROVED** [review]
